[DOC] Add Kafka Connect Build documentation (#4183)
strimzi/strimzi-kafka-operator: documentation/api/io.strimzi.api.kafka.model.connect.build.Build.adoc

Configures additional connectors for Kafka Connect deployments.

=== `output`

> In order to build new container images with additional connector plugins, Strimzi requires a container registry where the images can be pushed to, stored and pulled from.

Reviewer: ```suggestion
In order to build new container images with additional connector plugins, Strimzi requires a container registry where the images can be pushed to, stored, and pulled from.
```

> Strimzi does not run its own container registry, so users need to provide their own registry.

Reviewer: ```suggestion
Strimzi does not run its own container registry, so a registry must be provided.
```

> Strimzi is able to use private container registries as well as public registries such as link:https://quay.io/[Quay^] or link:https://hub.docker.com//[Docker Hub^].

Reviewer: ```suggestion
Strimzi supports private container registries as well as public registries such as link:https://quay.io/[Quay^] or link:https://hub.docker.com//[Docker Hub^].
```

The container registry is configured in the `.spec.build.output` section of the `KafkaConnect` custom resource.

> The `output` configuration supports two types: `docker` and `imagestream`.
> The `output` section is required.

Reviewer: ```suggestion
The `output` configuration, which is required, supports two types: `docker` and `imagestream`.
```

.Using Docker registry

To use a Docker registry, you have to specify the `type` as `docker`, and the `image` field with the full name of the new container image.

> The full name has to include:

Reviewer: ```suggestion
The full name must include:
```

> * The address of the registry.

Reviewer 2: Does the address include organization and image name? Can we make it clearer?

Author: I'm not sure we should mentioned them specifically since they are concepts used by Docker Hub or Quay ... but many registries do not use them. So I think it might be confusing.

> * The address of the registry.

Reviewer: ```suggestion
* The address of the registry
```

> * Port number if it is listening on non-standard port

Reviewer: ```suggestion
* Port number (if listening on a non-standard port)
```

* The tag of the new container image

> Examples of valid image names are:

Reviewer: ```suggestion
Example valid Docker image names:
```

* `docker.io/my-org/my-image/my-tag`
* `quay.io/my-org/my-image/my-tag`
* `image-registry.image-registry.svc:5000/myproject/kafka-connect-build:latest`

> Different Kafka Connect deployments should use a different image (for example at least with different tags).

Reviewer: ```suggestion
Each Kafka Connect deployment must use a separate image, which can mean different tags at the most basic level.
```

> If the registry requires authentication, the field `pushSecret` can be used to set a name of the Secret with the registry credentials.

Reviewer: ```suggestion
If the registry requires authentication, use the `pushSecret` to set a name of the Secret with the registry credentials.
```

> The Secret should use the `kubernetes.io/dockerconfigjson` type and contain the `.dockerconfigjson` file with the Docker credentials.

Reviewer: ```suggestion
For the Secret, use the `kubernetes.io/dockerconfigjson` type and a `.dockerconfigjson` file to contain the Docker credentials.
```

> For more details see link:https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#registry-secret-existing-credentials[Create a Secret based on existing Docker credentials^]

Reviewer: ```suggestion
For more information on pulling an image from a provate registry, see link:https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#registry-secret-existing-credentials[Create a Secret based on existing Docker credentials^].
```

Reviewer: Do we want this link in the attributes file? I guess it might depend on whether there's a downstream equivalent to swap. (Off topic - maybe we should look at how we manage external refs. Not sure of the necessity of using the attributes file if links don't change)

Author: I can move the external link to the attributes. I guess it makes sense since we do it for most other external links. But I do not think we need to replace this one for downstream. But having them in one place will make it easier to mass update them in the future for example etc.

[source,yaml,subs=attributes+,options="nowrap"]
.Example `output` configuration
----
apiVersion: {KafkaConnectApiVersion}
kind: KafkaConnect
metadata:
  name: my-connect-cluster
spec:
  #...
  build:
    output:
      type: docker # <1>
      image: my-registry.io/my-org/my-connect-cluster:latest # <2>
      pushSecret: my-registry-credentials # <3>
  #...
----

> <1> The type of the output which should be used by Strimzi. (Required)

Reviewer: ```suggestion
<1> The type of the output used by Strimzi. (Required)
```

> <2> The full name of the image which will be used including the repository and tag. (Required)

Reviewer: ```suggestion
<2> The full name of the image used, including the repository and tag. (Required)
```

> <3> Name of the secret with the container registry credentials. (Optional)

Reviewer: ```suggestion
<3> (Optional) Name of the secret with the container registry credentials. 
```

.Using OpenShift ImageStream

> Alternatively, you can also use OpenShift ImageStream to store the new container image.

Reviewer: ```suggestion
Instead of Docker, you can use OpenShift ImageStream to store the new container image.
```

> The ImageStream has to be created manually before deploying the Kafka Connect.

Reviewer: ```suggestion
The ImageStream has to be created manually before deploying Kafka Connect.
```

> To use OpenShift ImageStream, you have to set the `type` to `imagestream`, and the `image` field should contain the name of the ImageStream and the tag which should be used.

Reviewer: ```suggestion
To use ImageStream, set the `type` to `imagestream`, and use the `image` field to specify the name of the ImageStream and the tag used.
```

> For example `my-connect-image-stream:latest`.

Reviewer: ```suggestion
For example, `my-connect-image-stream:latest`.
```

[source,yaml,subs=attributes+,options="nowrap"]
.Example `output` configuration
----
apiVersion: {KafkaConnectApiVersion}
kind: KafkaConnect
metadata:
  name: my-connect-cluster
spec:
  #...
  build:
    output:
      type: imagestream # <1>

>       imageStream: image-stream-name # <2>

Reviewer 2: Let's try to use a real example as for docker

  #...
----

> <1> The type of the output which should be used by Strimzi. (Required)

Reviewer: ```suggestion
<1> (Required) The type of the output used by Strimzi. 
```

> <2> The name of the ImageStream and tag. (Required)

Reviewer: ```suggestion
<2> (Required) Name of the ImageStream and tag. 
```

=== `plugins`

> 

Reviewer: Connector plugins are a set of JAR files that define the implementation required to connect to certain types of external system.

> The connector plugins which will be added to the container image are configured in the `.spec.build.plugins` section of the `KafkaConnect` custom resource.

Reviewer: ```suggestion
The connector plugins added to a container image are configured using the `.spec.build.plugins` property of the `KafkaConnect` custom resource.
```

> Each connector plugin needs to have a name which is unique within the Kafka Connect deployment.

Reviewer: ```suggestion
Each connector plugin must have a name which is unique within the Kafka Connect deployment.
```

> Additionally, it contains a list of artifacts which it consists of.

Reviewer: ```suggestion
Additionally, the plugin artifacts must be listed.
```

> These artifacts will be downloaded by Strimzi, added to the new container image, and used in the Kafka Connect deployment.

Reviewer: ```suggestion
These artifacts are downloaded by Strimzi, added to the new container image, and used in the Kafka Connect deployment.
```

> The connector plugin artifacts can also include additional components such as (de)serializers and so on.

Reviewer: ```suggestion
The connector plugin artifacts can also include additional components, such as (de)serializers.
```

> Each connector plugin will be downloaded into a separate directory so that the different connectors and their dependencies are properly sand-boxed.

Reviewer: ```suggestion
Each connector plugin is downloaded into a separate directory so that the different connectors and their dependencies are properly _sandboxed_.
```

> The `plugins` section is required.

Reviewer: ```suggestion

```

Reviewer: We say it is required at the start of the section

> Each plugin has to be configured with at least one `artifact`.

Reviewer: ```suggestion
Each plugin must be configured with at least one `artifact`.
```

[source,yaml,subs=attributes+,options="nowrap"]
.Example `plugins` configuration with two connector plugins
----
apiVersion: {KafkaConnectApiVersion}
kind: KafkaConnect
metadata:
  name: my-connect-cluster
spec:
  #...
  build:
    output:
      #...
    plugins: # <1>
      - name: debezium-postgres-connector
        artifacts:
          - type: tgz
            url: https://repo1.maven.org/maven2/io/debezium/debezium-connector-postgres/1.3.1.Final/debezium-connector-postgres-1.3.1.Final-plugin.tar.gz

>             sha512sum: 962a12151bdf9a5a30627eebac739955a4fd95a08d373b86bdcea2b4d0c27dd6e1edd5cb548045e115e33a9e69b1b2a352bee24df035a0447cb820077af00c03

Reviewer: Can we snip these hashes?

Author: I was thinking about it. I think we have two options ... either use some completely artificial examples where we can keep some short URL and short (invalid) checksum. Or we can use areal connectors as we have there now. I do not think we should mix it - have a real URL with wrong checksum since that would confuse people who try to use it. 

I thought about the length already when writing it. But it seemed ok to me to go with the real thing. But I do not have strong opinions. So if you think I can change it to fake values (which I used anyway for the JAR artifact where I did not have a nice example - apart from my own connector which a bit seemed weird to use here).

Reviewer: I see. No strong opinions. It was really about the presentation, which is secondary if you think real examples might even be used. One potential issue on using real examples is maintenance, and how long we expect them to be relevant.

      - name: camel-telegram
        artifacts:
          - type: tgz
            url: https://repo.maven.apache.org/maven2/org/apache/camel/kafkaconnector/camel-telegram-kafka-connector/0.7.0/camel-telegram-kafka-connector-0.7.0-package.tar.gz
            sha512sum: a9b1ac63e3284bea7836d7d24d84208c49cdf5600070e6bd1535de654f6920b74ad950d51733e8020bf4187870699819f54ef5859c7846ee4081507f48873479
  #...
----

> <1> List of connector plugins and their artifacts which should be downloaded. (Required)

Reviewer: ```suggestion
<1> (Required) List of connector plugins and their artifacts, which must be downloaded. 
```

Reviewer: Don't think we need the download bit here, as it is mentioned next

> Currently, Strimzi supports two types of artifacts:

Reviewer: ```suggestion
Strimzi supports two types of artifacts:
```

> * Downloading JAR files and using them directly
> * Downloading TGZ archives and unpacking them

Reviewer: ```suggestion
* JAR files, which are downloaded and used directly
* TGZ archives, which are downloaded and unpacked
```

> IMPORTANT: Strimzi does not do any security scanning of the downloaded artifacts.

Reviewer: ```suggestion
IMPORTANT: Strimzi does not perform any security scanning of the downloaded artifacts.
```

> For security reasons, you should first verify the artifacts manually and configure the checksum verification to make sure the same artifact is used in the automated build and in the Kafka Connect deployment.

Reviewer: ```suggestion
For security reasons, you should first verify the artifacts manually, and configure the checksum verification to make sure the same artifact is used in the automated build and in the Kafka Connect deployment.
```

.Using JAR artifacts

> The JAR artifacts represent a resource which will be downloaded and added to the container image.

Reviewer: ```suggestion
JAR artifacts represent a resource which is downloaded and added to a container image.
```

> It is intended to be mainly used for downloading JAR files.
> It can be also used to download other file types.

Reviewer: ```suggestion
JAR artifacts are mainly used for downloading JAR files, but they can also used to download other file types.
```

> To use a JAR artifacts, you have to set the `type` field to `jar` and use the `url` to specify the URL from which it should be downloaded.

Reviewer: ```suggestion
To use a JAR artifacts, set the `type` property to `jar`, and specify the download location using the `url` property.
```

> Additionally, you can specify SHA-512 checksum of the file which will be downloaded.

Reviewer: ```suggestion
Additionally, you can specify a SHA-512 checksum of the artifact.
```

If specified, Strimzi will verify the checksum of the artifact while building the new container image.

[source,yaml,subs=attributes+,options="nowrap"]

> .Example of a JAR artifact

Reviewer: ```suggestion
.Example JAR artifact
```

----
apiVersion: {KafkaConnectApiVersion}
kind: KafkaConnect
metadata:
  name: my-connect-cluster
spec:
  #...
  build:
    output:
      #...
    plugins:
      - name: my-plugin
        artifacts:
          - type: jar # <1>
            url: https://my-domain.tld/my-jar.jar # <2>
            sha512sum: 589...ab4 # <3>
          - type: jar
            url: https://my-domain.tld/my-jar2.jar
  #...
----

> <1> Type of the artifact. (Required)

Reviewer: ```suggestion
<1> (Required) Type of artifact. 
```

> <2> The URL from which the artifact will be downloaded. (Required)

Reviewer: ```suggestion
<2> (Required) URL from which the artifact is downloaded. 
```

> <3> The SHA-512 checksum to verify the artifact. (Optional)

Reviewer: ```suggestion
<3> (Optional) The SHA-512 checksum to verify the artifact. 
```

.Using TGZ artifacts

> The TGZ artifacts can be used to download TAR archives compressed using the Gzip compression.

Reviewer: ```suggestion
TGZ artifacts are used to download TAR archives that have been compressed using Gzip compression.
```

> The TGZ archive can contain the whole Kafka Connect connector even if it consists of multiple different files.

Reviewer: ```suggestion
The TGZ artifact can contain the whole Kafka Connect connector, even when comprising multiple different files.
```

> It will be automatically downloaded and unpacked by Strimzi while building the new container image.

Reviewer: ```suggestion
The TGZ artifact is automatically downloaded and unpacked by Strimzi while building the new container image.
```

> To use a TGZ artifacts, you have to set the `type` field to `tgz` and use the `url` to specify the URL from which the TGZ archive should be downloaded.

Reviewer: ```suggestion
To use TGZ artifacts, set the `type` property to `tgz`, and specify the download location using the `url` property.
```

> Additionally, you can specify SHA-512 checksum of the archive which will be downloaded.

Reviewer: ```suggestion
Additionally, you can specify a SHA-512 checksum of the artifact.
```

If specified, Strimzi will verify the checksum before unpacking it and building the new container image.

[source,yaml,subs=attributes+,options="nowrap"]

> .Example of a JAR artifact

Reviewer 2: It's TGZ here

> .Example of a JAR artifact

Reviewer: ```suggestion
.Example TGZ artifact
```

----
apiVersion: {KafkaConnectApiVersion}
kind: KafkaConnect
metadata:
  name: my-connect-cluster
spec:
  #...
  build:
    output:
      #...
    plugins:
      - name: my-plugin
        artifacts:
          - type: tgz # <1>
            url: https://my-domain.tld/my-connector-archive.jar # <2>
            sha512sum: 158...jg10 # <3>
  #...
----

> <1> Type of the artifact. (Required)

Reviewer: ```suggestion
<1> (Required) Type of artifact. 
```

> <2> The URL from which the archive will be downloaded. (Required)

Reviewer: ```suggestion
<2> (Required) URL from which the archive is downloaded. 
```

> <3> The SHA-512 checksum to verify the archive. (Optional)

Reviewer: ```suggestion
<3> (Optional) SHA-512 checksum to verify the archive. 
```
